src/transformers/models/aimv2/modular_aimv2.py

```
@@ -0,0 +1,724 @@
+        )
+
+        last_hidden_state = encoder_outputs[0]
+        last_hidden_state = self.rms_norm(last_hidden_state)
+
+        pooler_output = self.head(last_hidden_state) if self.use_head else None
+
+        return BaseModelOutputWithPooling(
+            last_hidden_state=last_hidden_state,
+            pooler_output=pooler_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
+
+
+@auto_docstring(
+    custom_intro="""
+    The text model from AIMv2 without any head or projection on top.
+    """
+)
+class Aimv2TextModel(Aimv2PreTrainedModel):
+    main_input_name = "input_ids"
+
+    def __init__(self, config: Aimv2TextConfig):
+        super().__init__(config)
+        self.config = config
+        self.embeddings = Aimv2TextEmbeddings(config)
+        self.encoder = Aimv2Encoder(config)
+        self.rms_norm = Aimv2RMSNorm(config.hidden_size, config.rms_norm_eps)
+
+        self.eos_token_id = config.eos_token_id
+
+        self.post_init()
+
+    def get_input_embeddings(self) -> nn.Module:
+        return self.embeddings.token_embedding
+
+    def set_input_embeddings(self, value):
+        self.embeddings.token_embedding = value
+
+    @can_return_tuple
+    @auto_docstring
+    def forward(
+        self,
+        input_ids,
+        attention_mask: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+    ) -> BaseModelOutputWithPooling:
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+
+        hidden_states = self.embeddings(input_ids)
+        _, seq_len, _ = hidden_states.shape
+
+        if attention_mask is not None:
```

Here we should remove the check altogether (I think it was suppoosed to check the opposite anyway, but `create_causal_mask` will work independently of the mask being None or not)

---

@Cyrilvallez  Not quite working correctly, in eager implementation we are always returning a mask even when attention_mask is None whereas in sdpa implementation we are correctly returning None. AFAIK in eager mode we are creating the mask coz `allow_is_causal_skip` is False by default (so sdpa func returns a causal mask in this but in sdpa implementation that flag is True) :thinking: but tbh didn't completely get the new mask utils logic so I might wrong in my understanding.

---

Ha yes it will always return a mask because it creates a causal one! If the attention is not causal and we never use a mask, then we should remove that function altogether!